Version 7.0.0 release blog post (#178)
apache/arrow-site: _posts/2022-01-19-7.0.0-release.md

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
{% endcomment %}
-->

The Apache Arrow team is pleased to announce the 7.0.0 release. This covers
over 3 months of development work and includes [**474 resolved issues**][1]
from [**?? distinct contributors**][2]. See the Install Page to learn how to
get the libraries for your platform.

The release notes below are not exhaustive and only expose selected highlights
of the release. Many other bugfixes and improvements have been made: we refer
you to the [complete changelog][3].

## Community

Since the 6.0.1 release, Rémi Dattai and Alessandro Molina have been invited to be committers. 
Daniël Heres and Yibo Cai have joined the Project Management Committee (PMC). 
Thanks for your contributions and participation in the project!

## Columnar Format Notes

TBD

## Arrow Flight RPC notes

> TBD

Reviewer: ```suggestion
The Flight specification has been clarified to note that schemas are expected to be IPC-encapsulated on the wire.

Documentation has been generally improved; see the [Arrow Cookbook](https://arrow.apache.org/cookbook/) for recipes on how to use Flight in Python and R, and a new [example](https://github.com/apache/arrow/blob/master/cpp/examples/arrow/flight_grpc_example.cc) on how to use Flight and gRPC services on the same port.

This release includes Arrow Flight SQL, a protocol for using Arrow Flight to execute queries against and fetch metadata from SQL databases. Support is included for C++ and Java. A more detailed blog post is forthcoming. Note that development is ongoing and the specification is currently experimental.
```

Reviewer: CC @jduong @rafael-telles @vfraga if there's anything else you would like to add about Flight SQL here (of course we can have a blog post specifically about Flight SQL as well).

Reviewer 2: Does "support is included for C++" mean _just_ for C++, or has support also been integrated in bindings that use the C++ API, like Python?

Reviewer: @chrish42 in this case it means only C++, I'll update this to clarify.

## C++ notes

TBD

### Compute Layer

TBD

### Dataset Layer

TBD

> 

Reviewer: ```suggestion
This release includes [Skyhook](https://jayjeetc.medium.com/skyhookdm-is-now-a-part-of-apache-arrow-e5d7b9a810ba), a Dataset extension that can push down file format decoding, projection, and filtering to Ceph; this reduces client CPU bottlenecks, improves scan performance, and better utilizes storage cluster resources. A more detailed blog post about Skyhook is forthcoming.
```

Reviewer: CC @JayjeetAtGithub if you'd like to make any other notes about Skyhook.

### IO and Filesystem Layer

TBD

### Parquet

TBD

## C# notes

TBD

> ## Go notes
> 

Reviewer 3: ```suggestion
## Go notes

### Arrow

#### Bug Fixes

* License lifted up a level so that it is properly detected for the github.com/apache/arrow/go/v7 module for pkg.go.dev [ARROW-14728](https://github.com/apache/arrow/pull/11715). Documentation on pkg.go.dev will look correct with complete major version handling as of the v7.0.0 release.
* Errors from `MessageReader.Message` get properly surfaced by `Reader.Read` [ARROW-14769](https://github.com/apache/arrow/pull/11739)
* `ipc.Reader` properly uses the allocator it is initialized with instead of making native byte slices [ARROW-14717](https://github.com/apache/arrow/pull/11712)
* Fixed a CI issue where the CGO tests were crashing on windows [ARROW-14589](https://github.com/apache/arrow/pull/11611)
* Various fixes for internal usages of `Release` and `Retain` to maintain proper management of reference counting.

#### Enhancements

* Continuous Integration for Go library now uses Go1.16 as the version being tested [ARROW-14985](https://github.com/apache/arrow/pull/11860)
* `ValueOffsets` function added to `array.String` to return the entire slice of offsets [ARROW-14645](https://github.com/apache/arrow/pull/11653)
* `array.Interface` has been lifted to `arrow.Array`, `array.{Record,Column,Chunked,Table}` have been lifted to `arrow.{Record,Column,Chunked,Table}`. Interface `arrow.ArrayData` has been created to be used instead of `array.Data`. Aliases have been provided for the `array` package so existing code that doesn't directly use `array.Data` shouldn't be affected. The aliases will be removed in v8. [ARROW-5599](https://github.com/apache/arrow/pull/11832)
* Arrays and Records now support marshalling to JSON via the `json.Marshaller` interface. Builders support adding values to them by unmarshalling from JSON via the `json.Unmarshaller` interface. `array.FromJSON` function added to create Arrays from JSON directly. [ARROW-9630](https://github.com/apache/arrow/pull/11359)
* Basic handling of field referencing and expression building similar to the C++ Compute APIs added through the new `compute` package in preparation for adding compute interfaces. Does not yet allow *executing* expressions. [ARROW-14430](https://github.com/apache/arrow/pull/11514)

### Parquet

#### Enhancements

* Updated dependency versions [ARROW-14462](https://github.com/apache/arrow/pull/11537)
* `file` module added, Go Parquet library now supports full file reading and writing. [ARROW-13984](https://github.com/apache/arrow/pull/11146) [ARROW-13986](https://github.com/apache/arrow/pull/11538). Does not yet provide direct Parquet <--> Arrow conversions.
```

See the License for the specific language governing permissions and
limitations under the License.
{% endcomment %}
-->

The Apache Arrow team is pleased to announce the 7.0.0 release. This covers
over 3 months of development work and includes [**474 resolved issues**][1]
from [**?? distinct contributors**][2]. See the Install Page to learn how to
get the libraries for your platform.

The release notes below are not exhaustive and only expose selected highlights
of the release. Many other bugfixes and improvements have been made: we refer
you to the [complete changelog][3].

## Community

Since the 6.0.1 release, Rémi Dattai and Alessandro Molina have been invited to be committers.
Daniël Heres and Yibo Cai have joined the Project Management Committee (PMC).
Thanks for your contributions and participation in the project!

## Columnar Format Notes

TBD

## Arrow Flight RPC notes

TBD

## C++ notes

TBD

### Compute Layer

TBD

### Dataset Layer

TBD

### IO and Filesystem Layer

TBD

### Parquet

TBD

## C# notes

TBD

## Go notes

### Bug Fixes

TBD

### Enhancements

TBD

## Java notes

TBD

## JavaScript notes

TBD

## Python notes

TBD

## R notes

TBD

> ## Ruby and C GLib notes
> 
> ### Ruby
> 
> The updates of Red Arrow etc. consists of the following improvements:
> 
> TBD
> 
> ### C GLib
> 
> The updates of Arrow GLib etc. consists of the following improvements:
> 
> TBD
> 

Reviewer 4: ```suggestion
## Ruby and C GLib notes

### Ruby

There are two new contributors @okadakk and @simpl1g .

The updates of Red Arrow consists of the following improvements:
- Addition of a [cookbook section](https://github.com/apache/arrow/pull/11584) to the [main repository](https://github.com/apache/arrow/tree/master/ruby)
- Improved [Arrow::Function#execute useability](https://github.com/apache/arrow/pull/12101)
- Support for [table joins](https://github.com/apache/arrow/pull/12108)
- A [MacOS Development guide](https://github.com/apache/arrow/pull/11181)
- Enable separate use of [min and max](https://github.com/apache/arrow/pull/11179)
- Add support for loading [Arrow::Table from a URI](https://github.com/apache/arrow/pull/12128)

### C GLib

The updates of Arrow GLib etc. consists of the following improvements:

- [Rename GArrowSortKey::name to ::target](https://github.com/apache/arrow/pull/11748)
- Support for initializing [Amazon S3 APIs explicitly](https://github.com/apache/arrow/pull/11647)
- [Add support for expression](https://github.com/apache/arrow/pull/11304)
- [Add GArrowTableConcatenateOptions](https://github.com/apache/arrow/pull/11304)
- [Add support for filter](https://github.com/apache/arrow/pull/11442)

```

Reviewer 5: @bkmgit Thank you very much for your suggestion.  I'll complete the part of Ruby and C GLib based on yours.

Reviewer 4: @mrkn Thanks.

## Rust notes

> Rust releases minor versions every 2 weeks in addition to a major

Reviewer 6: FYI I added content for the 7.0.0 release of Rust -- fyi @jhorstmann @Dandandan @houqp @jorgecarleitao @andygrove @nevi-me

WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
{% endcomment %}
-->

The Apache Arrow team is pleased to announce the 7.0.0 release. This covers
over 3 months of development work and includes [**474 resolved issues**][1]
from [**?? distinct contributors**][2]. See the Install Page to learn how to
get the libraries for your platform.

The release notes below are not exhaustive and only expose selected highlights
of the release. Many other bugfixes and improvements have been made: we refer
you to the [complete changelog][3].

## Community

Since the 6.0.1 release, Rémi Dattai and Alessandro Molina have been invited to be committers.
Daniël Heres and Yibo Cai have joined the Project Management Committee (PMC).
Thanks for your contributions and participation in the project!

## Columnar Format Notes

TBD

## Arrow Flight RPC notes

TBD

## C++ notes

TBD

### Compute Layer

TBD

### Dataset Layer

TBD

### IO and Filesystem Layer

TBD

### Parquet

TBD

## C# notes

TBD

## Go notes

### Bug Fixes

TBD

### Enhancements

TBD

## Java notes

> TBD

Author: ```suggestion
* Flight SQL support is now available in the Java library, with integration tests to verify it against the C++ reference implementation.
* `GeneralOutOfPlaceVectorSorter` is now available for sorting any kind of vector. In general if dedicated sorters can be used (like `FixedWidthInPlaceVectorSorter`) they should preferred as they will generally perform better.
* `log4j2` dependency was removed as it was unused and a possible vector for attacks 
* `VectorSchemaRootAppender` now works with `BitVector`
```

## JavaScript notes

> TBD

Reviewer 7: ```suggestion
* Major simplifications to the API. There is only a single Vector class now. See the (also much improved) docs for details.
* Dictionary vectors created with `vectorFromArray` are automatically cached for better performance.
* Better tree shaking support. Some bundles can now be only a few kb.
```

## Python notes

> TBD

Reviewer 8: ```suggestion
* Official support for Python 3.6 has been dropped.
* `random` and `indices_nonzero` compute functions are now supported in Python
* `pyarrow.orc.read_table` is now provided to easily read the content of ORC files to a Table.
* `pyarrow.orc.ORCFile` now has a lot more properties exposed.
* `pyarrow.orc.ORCWriter` and `pyarrow.orc.write_table` now have numerous writer options available.
* `pyarrow.orc` now has much better API documentation.
* Support for compute functions arguments and options has been improved in general,  arguments are not position only, while options can be provided as keyword args or not, and error reporting for wrong arguments has been improved.
* `Table` now has a `group_by` method that allows to perform aggregations on table data. The compute functions documentation has also been improved to better distinguish between standard compute functions and `HASH_AGGREGATE` compute functions that can only be using for aggregations.
* Python documentation now provides interlinking for references to parameter types and return values, thus making far easier to navigate the documentation.
```

## R notes

> TBD

Reviewer 9: Honestly I'd just link to the R changelog (URL `[4]` already), compare to what we usually do

> TBD

Author: Changed to look like past releases

- "ceil_temporal", "floor_temporal", "round_temporal" to adjust temporal input
  to an integral multiple of a given unit (ARROW-14822);
- "year_month_day" to extract the calendar components of the input (ARROW-15032);
- "random" to general random floating-point values between 0 and 1 (ARROW-12404);
- "indices_nonzero" to return the indices in the input where there are
  non-zero, non-null values (ARROW-13035).

Decimal data is now supported as input of the arithmetic kernels
(ARROW-13130).

Dictionary data is now supported as input of the hash join execution node
(ARROW-14181).

Residual predicates have been implemented in the hash join node
(ARROW-13643).

The "list_parent_indices" function now always returns int64 data
regardless of the input type (ARROW-14592).

Month-day-nano interval data is now supported as input of the same functions
as other interval types (ARROW-13989).

### CSV

The CSV writer got additional configuration options:
- the string representation of null values (ARROW-14905);
- the quoting strategy: always / never / as needed (ARROW-14905);
- the end of line character(s) (ARROW-14907)

### Dataset Layer

> "SkyHook", a dataset addition that offloads fragment scan operations to a
> Ceph distributed storage, was contributed (ARROW-13607).

Reviewer: ```suggestion
[Skyhook](https://jayjeetc.medium.com/skyhookdm-is-now-a-part-of-apache-arrow-e5d7b9a810ba),
a dataset addition that offloads fragment scan operations to a
Ceph distributed storage cluster, was contributed (ARROW-13607).
A more detailed post about Skyhook is forthcoming.
```

Reviewer: Moved this suggestion from above since Antoine added detailed C++ notes.

The dataset writer now exposes options `min_rows_per_group` and
`max_rows_per_group` to control the size of row groups created (ARROW-14426).

### IO and Filesystem Layer

A critical bug in the AWS SDK for C++ that risks losing data in S3 multipart
uploads has been circumvented (ARROW-14523).

The Google Cloud Storage filesystem is now featureful enough to pass all
generic filesystem tests (ARROW-14924).

The OpenAppendStream method of filesystems has been un-deprecated; however,
it still cannot be implemented for all filesystem backends (ARROW-14969).

A new function `arrow::fs::ResolveS3BucketRegion` allows resolving the
region where a particular S3 bucket resides (ARROW-15165).

The S3 filesystem now sets the Content-Type of output files to
"application/octet-stream" (instead of "application/xml" previously)
if not explicitly specified by the caller (ARROW-15306).

### IPC

Fine-grained I/O (coalescing) is now enabled in the synchronous (ARROW-12683)
and asynchronous (ARROW-14577) IPC reader.

It is now possible to set the compression level when using LZ4 compression
(ARROW-9648).

> 

Reviewer 8: ### ORC

The ORC adapters have been significantly improved. A lot more properties of the ORC reader as well as ORC writer options are now available. Moreover API docs for both the ORC reader and the ORC writer have been generated.  (ARROW-11297)

### Parquet

DELTA_BYTE_ARRAY-encoded data can now be read from (but not written to)
bytearray columns in Parquet files (PARQUET-492).

## C# notes

TBD

## Go notes

### Bug Fixes

TBD

### Enhancements

TBD

## Java notes

TBD

## JavaScript notes

TBD

## Python notes

TBD

## R notes

TBD

## Ruby and C GLib notes

> ### Ruby
> 
> There are two new contributors @okadakk and @simpl1g .
> 
> The updates of Red Arrow consists of the following improvements:
> - Addition of a [cookbook section](https://github.com/apache/arrow/pull/11584) to the [main repository](https://github.com/apache/arrow/tree/master/ruby)
> - Improved [Arrow::Function#execute useability](https://github.com/apache/arrow/pull/12101)
> - Support for [table joins](https://github.com/apache/arrow/pull/12108)
> - A [MacOS Development guide](https://github.com/apache/arrow/pull/11181)
> - Enable separate use of [min and max](https://github.com/apache/arrow/pull/11179)
> - Add support for loading [Arrow::Table from a URI](https://github.com/apache/arrow/pull/12128)

Reviewer 5: ```suggestion
### Ruby

There are two new contributors @okadakk and @simpl1g .

The updates of Red Arrow consists of the following improvements:

- `Arrow::Function#execute` now accepts an instance of an `Arrow::Column` as its argument [(ARROW-14551)](https://issues.apache.org/jira/browse/ARROW-14551)
- `Arrow::TableLoader` now supports `.arrows` files to load [(ARROW-15356)](https://issues.apache.org/jira/browse/ARROW-15356)
- Add support for loading `Arrow::Table` from URI in `Arrow::TableLoader` [(ARROW-14562)](https://issues.apache.org/jira/browse/ARROW-14562)
- `Arrow::Table` now supports to join two tables [(ARROW-14531)](https://issues.apache.org/jira/browse/ARROW-14531)
- `Arrow::Function#execute` gets more easier to use than before [(ARROW-15274)](https://issues.apache.org/jira/browse/ARROW-15274)
- `Arrow::SortKey#name` has been renamed to `Arrow::SortKey#target` [(ARROW-14784)](https://issues.apache.org/jira/browse/ARROW-14784)
- Add Cookbook section to documentation [(ARROW-14636)](https://issues.apache.org/jira/browse/ARROW-14636)
- Support the explicit initialization of S3 API by the `Arrow.s3_initialize` method [(ARROW-14637)](https://issues.apache.org/jira/browse/ARROW-14637)
- On macOS, stop specifying the version of openssl package explicitly when building the extension library [(ARROW-14619)](https://issues.apache.org/jira/browse/ARROW-14619)
```

Reviewer 5: @kou Could you check this change?

Reviewer 10: Thanks!

> - Add support for loading `Arrow::Table` from URI in `Arrow::TableLoader` [(ARROW-14562)](https://issues.apache.org/jira/browse/ARROW-14562)

How about "Add support for `Arrow::Table.load(URI("https:///.../xxx.arrow"))`" or something? `Arrow::TableLoader` is an internal class. It's not for users.

Reviewer 5: Updated. Thank you.

### C GLib

The updates of Arrow GLib etc. consists of the following improvements:

> - [Rename GArrowSortKey::name to ::target](https://github.com/apache/arrow/pull/11748)
> - Support for initializing [Amazon S3 APIs explicitly](https://github.com/apache/arrow/pull/11647)
> - [Add support for expression](https://github.com/apache/arrow/pull/11304)
> - [Add GArrowTableConcatenateOptions](https://github.com/apache/arrow/pull/11304)
> - [Add support for filter](https://github.com/apache/arrow/pull/11442)

Reviewer 5: ```suggestion
- Add `garrow_execute_plan_build_hash_join_node` function, `GArrowHashJoinNodeOption`, and `GArrowJoinType` [(ARROW-15288)](https://issues.apache.org/jira/browse/ARROW-15288)
- Add `garrow_function_get_options_type` function [(ARROW-15273)](https://issues.apache.org/jira/browse/ARROW-15273)
- Add `garrow_function_get_default_options` function [(ARROW-15267)](https://issues.apache.org/jira/browse/ARROW-15267)
- Add `GArrowRoundToMultipleOptions` to customize the `round_to_multiple` function [(ARROW-15216)](https://issues.apache.org/jira/browse/ARROW-15216)
- Add `garrow_function_all` function to list up all the functions [(ARROW-15205)](https://issues.apache.org/jira/browse/ARROW-15205)
    - In addition, add `garrow_function_get_name`, `garrow_function_equal`, and `garrow_function_to_string` functions for convenience
- Add `GArrowRoundOptions` [(ARROW-15204)](https://issues.apache.org/jira/browse/ARROW-15204)
- Add `garrow_struct_scalar_get_value` function for converting a C++ scalar value to a GLib value [(ARROW-15203)](https://issues.apache.org/jira/browse/ARROW-15203)
- Add the following three interval data types [(ARROW-15134)](https://issues.apache.org/jira/browse/ARROW-15134)
    - `GArrowMonthIntervalDataType` for the interval with the month component
    - `GArrowDayTimeIntervalDataType` for the interval with the days and the milliseconds components
    - `GArrowMonthDayNanoIntervalDataType` for the interval with the months, the days, and the nanoseconds components
- Rename `GArrowSortKey::name` to `::target` [(ARROW-14784)](https://issues.apache.org/jira/browse/ARROW-14784)
- Support the explicit initialization of S3 API by the `garrow_s3_initialize` function [(ARROW-14637)](https://issues.apache.org/jira/browse/ARROW-14637)
- `garrow_decimal128_new_string` and `garrow_decimal256_new_string` now returns errors when they gets a invalid decimal string [(ARROW-14530)](https://issues.apache.org/jira/browse/ARROW-14530)
- `garrow_decimal128_data_type_new` and `garrow_decimal256_data_type_new` functions now validates the given precision [(ARROW-14529)](https://issues.apache.org/jira/browse/ARROW-14529)

```

Reviewer 5: @kou Could you check this change?

Reviewer 10: +1

> 

Reviewer 5: I'm editing this part.

Reviewer 5: done

---
layout: post
title: "Apache Arrow 7.0.0 Release"

> date: "2022-01-19 00:00:00 -0600"

Author: ```suggestion
date: "2022-02-9 00:00:00 -0600"
```

## Python notes

* Official support for Python 3.6 has been dropped.
* `random` and `indices_nonzero` compute functions are now supported in Python
* `pyarrow.orc.read_table` is now provided to easily read the content of ORC files to a Table.
* `pyarrow.orc.ORCFile` now has a lot more properties exposed.
* `pyarrow.orc.ORCWriter` and `pyarrow.orc.write_table` now have the writer options available.
* `pyarrow.orc` now has much better API documentation.
* Support for compute functions arguments and options has been improved in general,  arguments are not position only, while options can be provided as keyword args or not, and error reporting for wrong arguments has been improved.
* `Table` now has a `group_by` method that allows to perform aggregations on table data. The compute functions documentation has also been improved to better distinguish between standard compute functions and `HASH_AGGREGATE` compute functions that can only be using for aggregations.
* Python documentation now provides interlinking for references to parameter types and return values, thus making far easier to navigate the documentation.

## R notes

> This release adds additional improvements to the dyplr interface, to CSV support and to the C-Data interface to exchange data with other languages. For more details, see the [complete R changelog](https://arrow.apache.org/docs/r/news/).

Author: ```suggestion
This release adds additional improvements to the dyplr interface, to CSV support and to the C-Data interface to exchange data with other languages. For more details, see the [complete R changelog][4].
```
